### DISCUSSION TITLE
Auto Login Function

### DISCUSSION THREAD

**Discussion by kleineskraut** [cefsharp/CefSharp#4425] [Q&A]
Hello! I hope it's not an old topic but I have searched very much and tried very much but find no good solution:

i have built a function to auto fill via javascript to auto login to website. this works mostly good but sometimes unfortunately not e.g. here: https://geschaeftskunden.dhl.de/ The fields are filled but the page does not validate the values so i can't login. When I click on the button the page says to fill in both fields and clears them again. Does anyone have an idea for me?

Here is the code to fill the user field:
```vb
Dim js_username As String = "
var userNameField = document.querySelector(""input[type='text']"");
userNameField.value = '" & user & "';
userNameField.dispatchEvent(new Event('input', {bubbles: true}));
"
browserPAN.ExecuteScriptAsync(js_username)
```

**Comment by aimcosoftware** [comment]
Try creating an insertText event:

`new InputEvent('input',{inputType:'insertText',data: user})`

**Reply by kleineskraut** [reply]
thank you for your answer and I have tried many combinations but unfortunately had no success.

**Reply by aimcosoftware** [reply]
Yes I see now. Sorry, it has worked on all other sites, I have tried.

**Reply by aimcosoftware** [reply]
It looks like `DevToolsExtensions.GetDevToolsClient(Browser).Input.InsertTextAsync` will work.
The validation does use the input event, but I suspect they are checking the isTrusted flag.
You need to select the element first, as it only works on the focused element.
Something like:
```
js_username = "document.querySelector(""input[type='text']""); userNameField.value = '" & user & "';userNameField.select()"
browserPAN.ExecuteScriptAsync(js_username)
await DevToolsExtensions.GetDevToolsClient(browserPAN).Input.InsertTextAsync(User)
```

**Reply by amaitland** [reply]
> await DevToolsExtensions.GetDevToolsClient(browserPAN).Input.InsertTextAsync(User)

`DevToolsClient` implements `IDisposable`, please make sure to dispose of the instance when you have finished with the reference.

**Comment by amaitland** [comment]
An alternative to using `JavaScript` is to use [CefSharp.Dom](https://github.com/cefsharp/CefSharp.Dom#readme) which provides a `DOM API` built on top of the `DevTools API`. Here is a `C#` example, there are options for `type/click/move` to automate websites. It's a fully `async API`.

```c#
// Add using CefSharp.Dom; to get access to the
// CreateDevToolsContextAsync extension method
var devtoolsContext = await chromiumWebBrowser.CreateDevToolsContextAsync();

var htmlInputElement = await devToolsContext.QuerySelectorAsync<HtmlInputElement>("input[type='text']");

//Type text in an input field
await htmlInputElement.TypeAsync("username");

// Manually dispose of context (prefer DisposeAsync as the whole API is async)
await devToolsContext.DisposeAsync();
```

**Comment by kleineskraut** [comment]
thanks for the help and solutions. With both I do not get on this particular page. Everywhere else it works without problems. You seem to be right with the isTrusted flag and they try to prevent exactly that.

**Comment by augbel72** [comment]
Hi Kleineskraut, were you able to solve it? I think I have the same problem as you accessing https://www.sisal.it/
Both fields populate correctly but the button remains disabled! I'm going crazy... or rather, I think I've already gone crazy!

                    Dim js_username As String = "
var userNameField = document.querySelector(""input[type='text']"");
userNameField.value = '" & strCurrUsername & "';
userNameField.dispatchEvent(new Event('input', {bubbles: true}));
"
                    CefBrowser.ExecuteScriptAsync(js_username)

                    Dim js_password As String = "
var passwordField = document.querySelector(""input[type='password']"");
passwordField.value = '" & strCurrPassword & "';
passwordField.dispatchEvent(new Event('input', {bubbles: true}));
"
                    CefBrowser.ExecuteScriptAsync(js_password)

**Reply by aimcosoftware** [reply]
It will be because the validation is looking for input with `isTrusted = true` the only way I've found round it is to add a space and use `Sendkeys.Send("{bs}")` from .Net.

**Comment by augbel72** [comment]
I give up! I've tried them all but I really don't understand how to do it... the "Login" button always remains disabled!
this is the link (assuming you want to try)...

https://areaprivata.sisal.it/loginJwt

if you ever succeed, please tell me how you did it, I can't. Ciao, I'm really very frustrated!

**Comment by kleineskraut** [comment]
I have just tried out something again and followed up on aimcosoftware's idea. This approach works with your site:

```
    Private Async Sub EnsureFocusOnField_ANDFILL(ByVal value As String, ByVal field As String)
        Dim isFocused As Boolean = Await CheckFocusAndRetrieveValue(field)
        While Not isFocused
            SendKeys.Send("{TAB}")
            Await Task.Delay(100)
            isFocused = Await CheckFocusAndRetrieveValue(field)
        End While
        SendKeys.Send(value)
    End Sub

    Private Async Function CheckFocusAndRetrieveValue(ByVal field As String) As Task(Of Boolean)
        Dim script As String = "
    (function() {
        var fieldname = document.querySelector(""input[type='" & field & "']"");
        return document.activeElement === fieldname; // Gibt true oder false zurück
    })();"

        Dim response = Await browserPAN.EvaluateScriptAsync(script)
        If response.Success AndAlso response.Result IsNot Nothing Then
            Dim isFocused As Boolean = Convert.ToBoolean(response.Result)
            Console.WriteLine($"Fokus im fieldname: {isFocused}")
            Return isFocused
        Else
            Console.WriteLine("Script-Ausführung fehlgeschlagen oder das Ergebnis ist null.")
            Return False
        End If
    End Function
```
You can call this function with:

```
        EnsureFocusOnField_ANDFILL(user, "text")
        EnsureFocusOnField_ANDFILL(pas, "password")
```

i don't think the solution is particularly good but i just don't have a better idea and have already given up

_________________________________________________________________________________

I am sorry but I have never found a solution. I have just tried again with your site but unfortunately have no success

```
Dim js_username As String = "
var userNameField = document.querySelector(""input[type='text']"");
if (userNameField !== null) {
    userNameField.value = '" & user & " ';
    userNameField.dispatchEvent(new Event('input', {bubbles: true}));
    userNameField.focus();
}
"
        Await ExecuteScriptAsyncWithDelay(js_username, 500) ' Warte 500ms

    Private Async Function ExecuteScriptAsyncWithDelay(script As String, delay As Integer) As System.Threading.Tasks.Task
        browserPAN.ExecuteScriptAsync(script)
        Await System.Threading.Tasks.Task.Delay(delay) ' Warte für n Millisekunden
        SendKeys.Send("{bs}") ' Entfernt das zuletzt hinzugefügte Leerzeichen
    End Function
```

**Comment by augbel72** [comment]
This (see below) WORKS!!!
thank you so much, you solved a problem that hadn't let me sleep for days... you were really very courteous (capable and prepared), thanks again! Ciao, Augusto

            EnsureFocusOnField_ANDFILL(strCurrUsername, "text")
            Pausa(1000) 'wait for 1000 ms
            EnsureFocusOnField_ANDFILL(strCurrPassword, "password")
            Pausa(1000) 'wait for 1000 ms

            Dim script_click = "(function (){
        document.getElementById('buttonAuth').click();
	})();"
            Dim response_click = Await CefBrowser.EvaluateScriptAsync(script_click)
